### PR TITLE
add auto accelerator

### LINKED ISSUE #41: enable `accelerator=auto`

### PR DESCRIPTION
**Pull request by aniketmaurya**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #41 

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/litserve/connector.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import sys
```
Comment (williamFalcon) [inline src/litserve/connector.py:1]: when you write a new file please add the copyright notice @aniketmaurya

File: src/litserve/connector.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                return self._choose_gpu_accelerator_backend()
+            except RuntimeError:
+                print("No GPU accelerator found, will default to cpu")
+        return "cpu"
+
+    @staticmethod
+    def _choose_gpu_accelerator_backend():
+        import torch
+
+        if check_cuda_with_nvidia_smi():
```
Comment (aniketmaurya) [inline src/litserve/connector.py:62]: `torch.cuda.is_available()` leads to multiprocessing error and recommends using `multiprocessing.set_start_method(‘spawn’)` which doesn't seem to work for Pipes on Linux.

File: src/litserve/connector.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+        if accelerator not in ["auto", "cpu", "cuda", "gpu", None]:
+            raise ValueError("accelerator must be one of 'auto', 'cpu', 'cuda', or 'gpu'")
+
+        if accelerator is None:
+            return "auto"
+        return accelerator
+
+    def _choose_auto_accelerator(self):
+        if "torch" in sys.modules:
```
Comment (aniketmaurya) [inline src/litserve/connector.py:51]: @lantiga what do you think about having other frameworks here?

**Review by lantiga — CHANGES_REQUESTED** [review]
Looks good, a couple of points to address

File: tests/test_lit_server.py [hunk 4]
```diff
@@ -183,3 +188,64 @@ def test_litapi_with_stream(simple_litapi):
...
+        pytest.param(
+            None,
+            "mps",
+            marks=pytest.mark.skipif(not torch.backends.mps.is_available(), reason="Only tested on Apple MPS"),
+        ),
+    ],
+)
+def test_auto_accelerator(input_accelerator, expected_accelerator):
+    server = LitServer(SimpleLitAPI(), devices=1, timeout=10, accelerator=input_accelerator)
+    assert server._connector.accelerator == expected_accelerator
```
[the remark this answers is not in the document]
Comment (aniketmaurya) [reply tests/test_lit_server.py:257]: added - https://github.com/Lightning-AI/litserve/pull/44/files#diff-274978086d8b4f049cffd0a53c0a2b1634c323f7ae581e87427c3b0c9046716aR260

File: src/litserve/connector.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            try:
+                return self._choose_gpu_accelerator_backend()
+            except RuntimeError:
+                print("No GPU accelerator found, will default to cpu")
```
[the remark this answers is not in the document]
Comment (aniketmaurya) [reply src/litserve/connector.py]: the exception was handled by defaulting to CPU when no GPU device were available. I have updated it to work without exception handling mechanism.

**Review by lantiga — APPROVED** [review]
Looks great!

We'll see if it's worth keeping the `Connector` abstraction in the future or if we can just flatten the logic down into the server. It may get more complicated in the future, in which case I'd keep it.

**Pull request merged**